### PR TITLE
API: Update difficulty attribute

### PR DESCRIPTION
**Pull request by Ali-Kabbadj**
Added the necessary validation attributes to the Difficulty property

## Summary

Added  [Range] and [Required] attributes to the Difficulty properly  in :
*CreateQuestionRequest

Added [Range] attributes to the Difficulty properly  in :
*UpdateQuuestionRequest

Added a custume Validation Attribute to the Difficulties properly  in : 
*GetQuestionRequest

Note: For Difficulties properly in GetQuesttionRequest , i added an extension to the ValidationAtribute Class to be able to validate the range of an IEnumerable< int > Difficulties

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Refactoring (non-breaking change to code and structure)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Notes

I was wondering if the custume validation for  IEnumerable<int> Difficulties should maybe return all the questions for the difficulties that were correct, and just ignore the out of range ones, instead of what my implementation does ,which is return an error if one of the values in the IEnumerable is out oof range.

Also , for some reason when updating a question ,if one of the values is not provided in the put request the value is set to null , i was wondering is that was by choice .

If there is anything to change or modify just let me know , thanks in advance!

### CODE REVIEW

File: src/WebApi/WebApi/Extensions/RangeValidationAttributeExtension.cs
Comment (petrspelos) [inline src/WebApi/WebApi/Extensions/RangeValidationAttributeExtension.cs:26]: `optional/suggestion`

I suppose this might read a little better:
```suggestion
                return !list.Any(i => i < Min || i > Max);
```

what do you think? 🤔

File: src/WebApi/WebApi/UseCases/v1/Question/CreateQuestion/CreateQuestionRequest.cs
Comment (petrspelos) [inline src/WebApi/WebApi/UseCases/v1/Question/CreateQuestion/CreateQuestionRequest.cs:13]: `minor`

Since we don't seem to use the `ErrorMessage` anywhere else, I suppose we can skip it here. 🤔 

I kinda also think the default behavior generates basically this exact message. Not a 100% on that, though. ☺️ 

```suggestion
    [Required]
```

File: src/WebApi/WebApi/UseCases/v1/Question/UpdateQuestion/UpdateQuestionRequest.cs
Comment (petrspelos) [inline src/WebApi/WebApi/UseCases/v1/Question/UpdateQuestion/UpdateQuestionRequest.cs:2]: Is this using directive needed? 🤔 
I'm not sure what CodeAnalysis brings to this file 😅

**Review by petrspelos — CHANGES_REQUESTED** [review]
Hey @Ali-Kabbadj!

Thank you so much for contributing. ☺️

There's a couple of super minor things for us to look at.

---

Either way, we're working on having some more issues available for the public so if you like the codebase feel free to stick around. 😉

File: src/WebApi/WebApi/Extensions/RangeValidationAttributeExtension.cs
Comment (Ali-Kabbadj) [reply src/WebApi/WebApi/Extensions/RangeValidationAttributeExtension.cs:26]: Yes , definitely better.
I'll change it to that .

File: src/WebApi/WebApi/Extensions/RangeValidationAttributeExtension.cs
Comment (TomasNiemczyk98) [inline src/WebApi/WebApi/Extensions/RangeValidationAttributeExtension.cs:20]: I would recommend using here `Any()` rather than `Count() > 0`, so we don't have to enumerate the collection just to check if it contains any elements.

```suggestion
            if (value is IEnumerable<int> list && list.Any())
```

File: src/WebApi/WebApi/UseCases/v1/Question/UpdateQuestion/UpdateQuestionRequest.cs
Comment (Ali-Kabbadj) [reply src/WebApi/WebApi/UseCases/v1/Question/UpdateQuestion/UpdateQuestionRequest.cs:2]: Both test failed , fantastic 😅😅
This is my first time contribution , so I'm sure i have done many thing the wrong way , i will follow the changed to the code you've suggested above , and also try and find where the problem here is

**Review by TomasNiemczyk98 — CHANGES_REQUESTED** [review]
Hey @Ali-Kabbadj,
like Petr said also welcome and thanks for your contribution, overall it looks really solid!

Side note: some of the documentation is not in a perfect state right now, so if there is anything not clear feel free to reach out and we would be glad to help you 🙂

File: src/WebApi/WebApi/UseCases/v1/Question/UpdateQuestion/UpdateQuestionRequest.cs
Comment (petrspelos) [reply src/WebApi/WebApi/UseCases/v1/Question/UpdateQuestion/UpdateQuestionRequest.cs:2]: Thank you, @Ali-Kabbadj! Don't worry, this is a good place to learn. 😅

Comment (Ali-Kabbadj) [reply src/WebApi/WebApi/UseCases/v1/Question/UpdateQuestion/UpdateQuestionRequest.cs:2]: @petrspelos Thanks a lot , i definitely have a lot of learning to do  !!

File: src/WebApi/WebApi/Extensions/RangeValidationAttributeExtension.cs
Comment (Ali-Kabbadj) [reply src/WebApi/WebApi/Extensions/RangeValidationAttributeExtension.cs:20]: @TomasNiemczyk98 Thanks you, much appreciated , will do !!

**Comment by TomasNiemczyk98** [comment]
Looking at the reason the pipeline is failing and it's due to a test that is not passing, the test is in `End2EndTests/QuestionTests/UpdateQuestionTest`, the test basically tries to change the difficulty of a question to some invalid value (previously this was 6), now this test will probably have to be changed a little bit since the validation is now performed on the controller.

`[Fact]
public async Task Update_InvalidRequest_ReturnsBadRequest()
{
    var request = new UpdateQuestionRequest
    {
        Difficulty = 6
    };

    var result = await EndpointCall(1, request);

    Assert.Equal(HttpStatusCode.BadRequest, result.StatusCode);
}`

File: src/WebApi/WebApi/UseCases/v1/Question/UpdateQuestion/UpdateQuestionRequest.cs
Comment (petrspelos) [inline src/WebApi/WebApi/UseCases/v1/Question/UpdateQuestion/UpdateQuestionRequest.cs:11]: @Ali-Kabbadj you're also failing a test here, since this **used to be** 1 - 5 and you made it 0 - 100

```suggestion
    [Range(1, 5)]
```

So with this one, I'd just put it back

**Review by petrspelos — CHANGES_REQUESTED** [review]

File: src/WebApi/WebApi/UseCases/v1/Question/UpdateQuestion/UpdateQuestionRequest.cs
Comment (petrspelos) [reply src/WebApi/WebApi/UseCases/v1/Question/UpdateQuestion/UpdateQuestionRequest.cs]: @TomasNiemczyk98 what does that really mean in that case? What should really be the range here?

Comment (TomasNiemczyk98) [reply src/WebApi/WebApi/UseCases/v1/Question/UpdateQuestion/UpdateQuestionRequest.cs]: The range should be the accepted values for our difficulty field, it was my understanding that we are switching from a range of 1-5 to a range of 0-100

**Comment by Ali-Kabbadj** [comment]
@TomasNiemczyk98 @petrspelos , why did we have to change the range  attribute  instead of the way the test works , or is that just temporary ?

**Comment by petrspelos** [comment]
@Ali-Kabbadj you're right, we should do that. I was just a little confused since you didn't link the issue with this PR.

Yeah, let's change the value in the test. ☺️

**Comment by TomasNiemczyk98** [comment]
@Ali-Kabbadj you were correct to change the range from 1-5 to 0-100 since that is the description of the task. The issue is that this will break our test, so that will need to be fixed (as mentioned above) before we can merge 🙂

**Comment by Ali-Kabbadj** [comment]
@petrspelos My bad , will do next time ,thanks

**Comment by petrspelos** [comment]
@Ali-Kabbadj just one should be enough. You can (if you're willing) add one that is too low and one that is too high. 🙂

**Comment by Ali-Kabbadj** [comment]
@petrspelos sorry for the caused trouble , really!

 i don't know how to read the log of the tests to understand where the problem has accured  .

Maybe in `UpdateQuestionUseCaseTest` in `UnitTests`? the instance of the  `QuestionModel`  has a `Difficulty=1` , and it is used in the method `Execute_PassValidInput_CallsInvalid()` , im not sure it's just a guess, i might have done something wrong again, probably 😅😅.

**Comment by TomasNiemczyk98** [comment]
@Ali-Kabbadj Hey, the fault is on our side, the workflow logs are pretty useless right now since they do not give any relevant information about what went wrong. I am guessing that the error is related to sonarcloud and it's secret key, but will have to verify tomorrow, I will let you know ASAP.

**Comment by Ali-Kabbadj** [comment]
Great . Thanks @TomasNiemczyk98 😊

**Comment by petrspelos** [comment]
@TomasNiemczyk98 actually @PlesnikJakub will merge this anyways this evening if everything turns out to be okay. It's just that we updated the pipeline while the PR is in progress. The best way forward is to wrap this up ASAP. ☺️

**Comment by PlesnikJakub** [comment]
Hey @Ali-Kabbadj, thank you for your contribution and I am sorry for the issues with the pipeline. 
Merging this now 🚀 

For context, it's really as other guys mentioned, since PR is based on fork pipeline doesn't have access to secrets required to execute sonar cloud analysis.  We will sort this out ASAP.

**Pull request merged**
